src/Workspaces/Core/Portable/Shared/Utilities/AsyncBatchingWorkQueue`2.cs

```diff
@@ -122,11 +108,35 @@ public void AddWork(IEnumerable<TItem> items)
             {
                 // add our work to the set we'll process in the next batch.
                 AddItemsToBatch(items);
-                TryKickOffNextBatchTask();
+
+                if (!_taskInFlight)
+                {
+                    // No in-flight task.  Kick one off to process these messages a second from now.
+                    // We always attach the task to the previous one so that notifications to the ui
+                    // follow the same order as the notification the OOP server sent to us.
+                    _updateTask = ContinueAfterDelay(_updateTask);
+                    _taskInFlight = true;
```

Comment:
💭 This is still problematic, but on other paths.

1. `await lastTask.ConfigureAwait(false)` will set the value to false, which could either be before or after the next item in the chain sets it to true
2. Even if (1) doesn't cause a problem, the fact that `ContinueAfterDelay` runs on a different thread means that it could reach the point of `ProcessNextBatchAsync` and still set the value to false before it gets set to true (a less likely but still possible equivalent to the synchronous completion case)

Reply:
➡️ Since `GetNextBatchAndResetQueue()` takes the same lock that this thread is using, we don't have to worry about these conditions.

Reply:
taken offline.  we're satisfied with this due to this code running in a lock.